### PR TITLE
Feature/dex 576 apm abstraction logic

### PR DESCRIPTION
**Pull request by EvansPie**
DEX-576

# What this PR does

Implements and consumes payment method config view model.

# Notable decisions & other stuff

List & explain ...

# Instructions on how to test this

Let other reviewers know how they can test this.

# Before merging

_QA_

- [x] I manually tested this with a demo application (simulator)
- [ ] I manually tested this with a demo application (real device)
- [ ] I wrote unit tests for each new util-like function
- [ ] I wrote at least basic e2e tests for the new functionalities or new paths
- [ ] If this PR introduces UI changes, I manually tested this PR on iPhone 8 (iOS 10, 12, 14), iPhone 12, and iPhone 12 Max
- [ ] If this PR modifies the behavior of an API, I did my best to make my new API backward compatible
- [ ] If this PR modifies the API, I attempted to implement the feature with an example

_Documentation_

- [ ] If this PR modifies the API, I updated the API Reference.
- [ ] If this PR adds new options to the API, I wrote some documentation or guide in the online documentation
- [ ] If the PR modifies the API, I wrote some documentation to deprecate it

# After merging

- Make sure a new release has been pushed to Cocoapods trunk

### CODE REVIEW

File: Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    }
+    
+    init(config: PaymentMethodConfig) {
+        self.config = config
+    }
+    
+    @objc
+    func buttonTapped() {
+        switch config.type {
+        case .hoolah,
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift:37]: when it comes to switching between all the payment methods, maybe we could add enum methods to reduce the number of touch points for every new payment method? We did a similar change to Android.

File: Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+                    Primer.shared.delegate?.checkoutFailed?(with: err)
+                    self.tokenizationCompletion?(nil, err)
+                    self.tokenizationCompletion = nil
+                }
+            }
+        }
+    }
+    
+    lazy var title: String = {
+        switch config.type {
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift:118]: same as above

Comment (ceriksson) [reply Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift:118]: alternatively we put this in global state alongside all the other copy

File: Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+                .googlePay,
+                .hoolah,
+                .payNLIdeal,
+                .payPal,
+                .klarna:
+            return nil
+        }
+    }()
+    
+    lazy var buttonBorderWidth: CGFloat = {
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/AsyncPaymentMethodTokenizationViewModel.swift:257]: this type of logic could be in its own descriptor class targeting payment method button component.

File: Sources/PrimerSDK/Classes/Data Models/PaymentMethodConfig.swift [hunk 4]
```diff
@@ -17,6 +17,10 @@ struct PrimerConfiguration: Codable {
             }
         }
         
+        for (index, viewModel) in viewModels.enumerated() {
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/PaymentMethodConfig.swift:20]: what's this doing?

File: Sources/PrimerSDK/Classes/Data Models/PrimerFlowEnums.swift [hunk 5]
```diff
@@ -81,15 +90,18 @@ internal enum PrimerInternalSessionFlow {
 
     case vault
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/PrimerFlowEnums.swift:91]: could we refactor this pattern somehow? starting to reach its limit

File: Sources/PrimerSDK/Classes/User Interface/Root/PrimerRootViewController.swift [hunk 6]
```diff
@@ -137,7 +136,8 @@ internal class PrimerRootViewController: PrimerViewController {
                 case .completeDirectCheckout:
                     break
                     
-                case .addPayPalToVault:
+                case .addPayPalToVault,
+                        .checkoutWithPayPal:
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/User Interface/Root/PrimerRootViewController.swift:140]: will this vault?

File: Sources/PrimerSDK/Classes/User Interface/Root/PrimerRootViewController.swift [hunk 7]
```diff
@@ -171,11 +171,35 @@ internal class PrimerRootViewController: PrimerViewController {
                 case .addApayaToVault:
                     self?.presentApaya()
                     
+                case .checkoutWithHoolah:
+                    guard let viewModel = PrimerConfiguration.paymentMethodConfigViewModels.filter({ $0.config.type == .hoolah }).first else { return }
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/User Interface/Root/PrimerRootViewController.swift:175]: Is there a way to refactor these repeated code snippets?

File: Sources/PrimerSDK/Classes/Data Models/PrimerFlowEnums.swift (hunk 5 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Data Models/PrimerFlowEnums.swift:91]: On my plans, this will be removed.

**Pull request merged**